[hail] improve makefile in anticipation of approximate quantiles

danking opened the pull request:
- automatically download catch.hpp from GitHub if it's missing
 - remove builtin rules and suffix based rules to improve performance and ease debugging
 - change Makefile variable definitions to match our standard `FOO OP VAL` (note: two spaces)
 - rely on `clang -MM` (see: [Clang command line reference](https://clang.llvm.org/docs/ClangCommandLineReference.html) and below explanation) to generate precise dependencies for object files (including source and header files)
 - explicitly specify which files depend on `NUMBER_OF_GENOTYPES_PER_ROW` to be set (namely the dependency file and the object file associated with `ibs.cpp`)
 - break `TEST_OBJECTS` into two steps so that we can have `_test.cpp` files which do not have corresponding `.cpp` files (consider, for example, a header-only file, which ApproximateQuantiles will be)
 - eliminate `$(BUILD)/headers` in favor of precise dependency tracking described above
 - remove the target `$(BUILD)`, directories don't work the way you think in Make, it's better to have individual rules create the containing directories when necessary
 - remove `wget` nonsense, standardize on `curl -sSL` (which produces useful error messages)

---

# clang -MM

This argument to clang allows us to generate "depfile" or "dependency files" which are valid `Makefile`s describing how object files depend on `c`, `cpp`, `h`, and `hpp` files.

`clang -MM foo.cpp` writes to stdout a Makefile that indicates how `foo.o` depends on preprocessor includes of other *user* files. For example,

```
# cat foo.cpp
#include<stdio.h>
#include "bar.h"
# clang -MM foo.cpp
foo.o: foo.cpp bar.h
```

The `-MT target` allows us to specify the target's name:
```
# clang -MM foo.cpp -MT fiddle
fiddle: foo.cpp bar.h
```

The `-MQ` argument asks `clang` to quote the variable before make sees it, so (nb, I first quote it for the shell so it doesn't get seen as an env var):
```
# clang -MM foo.cpp -MQ '$fiddle'
$$fiddle: foo.cpp bar.h
```

The `-MG` argument tells `clang` to not error if an included file does not exist. This can be helpful if the project's `Makefile` knows how to generate these header files. In our example, `catch.hpp` can be generated by `hail/src/main/c/Makefile` by downloading it from GitHub.
```
# rm -rf bar.h
# clang -MM foo.cpp              
foo.cpp:1:10: fatal error: 'bar.h' file not found
#include "bar.h"
         ^~~~~~~
1 error generated.
# clang -MM foo.cpp -MG
foo.o: foo.cpp bar.h
```

The `-MF file` argument tells `clang` to write to `file` instead of stdout.

The `-MMD` argument is used if we *also want to compile the file*. The `-MM` argument defaults to adding `-E` meaning "only run the preprocessor".

danking commented:
I assigned @catoverdrive, but @cseed @tpoterba @chrisvittal may also be interested and/or have useful feedback.

catoverdrive commented:
I've looked through this and it looks fine to me. I'll leave approval to @chrisvittal

danking commented:
A wee little bump, I'm curious to hear more about the build issue in unclean directories!

chrisvittal commented:
After looking more at the build logs here, as I was curious as to how the tests passed, when `make` was run in compilation, it ran the target for `simd/simd.h`, which ran the `$(LIBSIMDPP)` target, which ran `tar xzf $(LIBSIMDPP).tar.gz`. Thus accidentally creating the necessary dependencies before the `test` target is run later in the CI build.

danking commented:
@chrisvittal I think this simd issue may have been the clang/g++ issue you noted?

danking commented:
Response:

1. `simdpp/simd.h`, I changed to a user dependency *and* changed the makefile to grab system dependencies, I agree this seems like a good idea (but I'm not sure, so we'll try it and find out!)
2. `mkdir -p`, I changed them all to `@`, but can `mkdir -p` fail (except for, say, filesystem being full)?
3. `.DEFAULT_GOAL` good catch, I think we don't want `.DEFAULT` at all, I've removed it

chrisvittal commented on hail/src/main/c/Makefile:
```diff
@@ -1,11 +1,12 @@
+MAKEFLAGS += --no-builtin-rules
+.PHONY: all clean debug prebuilt test
+.SUFFIXES:
+.DEFAULT_GOAL: all
```
I know this is wierd, but `.DEFAULT_GOAL` is defined as a variable, not a target.
```make
.DEFAULT_GOAL := all
```

chrisvittal commented on hail/src/main/c/Makefile:
```diff
@@ -39,14 +42,33 @@ OBJECTS := \
+	@mkdir -p $(@D)
+	$(CXX) -o $@ $(CXXFLAGS) -MD -MF $(@:%.o=%.d) -MT $@ -c $<
+
+$(BUILD)/ibs.d: CXXFLAGS += -DNUMBER_OF_GENOTYPES_PER_ROW=256
```
More GCC problems! It appears that in the preprocessor for gcc < 8, it treats `(X % 0) != 0` as true‽ And still respects `#error` even in the presence of -MG.

```
$ CXX=g++-7 make build/ibs.d
g++-7 -march=sandybridge -O3 -std=c++14 -Ilibsimdpp-2.1 -Wall -Werror -fPIC -ggdb -fno-strict-aliasing -I../resources/include -I/usr/lib/jvm/java-8-openjdk/include -I/usr/lib/jvm/java-8-openjdk/include/linux -DNUMBER_OF_GENOTYPES_PER_ROW=256 ibs.cpp -MG -M -MF build/ibs.d -MT build/ibs.o
In file included from ibs.cpp:5:0:
ibs.h:36:46: error: division by zero in #if
 #if (NUMBER_OF_UINT64_GENOTYPE_PACKS_PER_ROW % UINT64_VECTOR_SIZE) != 0
                                              ^
ibs.h:37:2: error: #error "genotype packs per row, NUMBER_OF_UINT64_GENOTYPE_PACKS_PER_ROW, must be multuple of vector width, UINT64_VECTOR_SIZE."
 #error "genotype packs per row, NUMBER_OF_UINT64_GENOTYPE_PACKS_PER_ROW, must be multuple of vector width, UINT64_VECTOR_SIZE."
  ^~~~~
```
Potential workaround, add `-DHAIL_OVERRIDE_WIDTH=4` to just this rule

danking replied:
oy vey

danking replied:
I'm sorry I wrote `ibs.cpp` this is all my fault.
